shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -272,16 +297,131 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+        thisCtx.targsMaps += td.nme.name -> MutMap("" -> targsMap.collect { case s -> (tv: TypeVariable) =>
+          val rigid = freshenAbove(ctx.lvl, tv, true)
+          m += rigid -> tv
+          s -> rigid
+        })
+        val reverseRigid = m.toMap
+        val thisTy = TypeRef(td, td.tparams.map(p => thisCtx.targsMaps(tn.name)("")(p.name)))(prov, thisCtx)
+        thisCtx += "this" -> thisTy
+        td.mthDecls.foreach { case md @ MethodDef(rec, prt, nme, tparams, R(ty)) =>
+          implicit val prov: TypeProvenance = tp(md.toLoc, "method declaration")
+          val dummyTargs2 = tparams.map(p => freshVar(noProv/*TODO*/)(ctx.lvl + 2))
+          allTargsMaps(tn.name) += nme.name -> tparams.map(_.name).zip(dummyTargs2).toMap
+          if (!nme.name.head.isUpper)
+            err(msg"Method names must start with a capital letter", md.toLoc)
+          val bodyTy = subst(typeType(ty)(thisCtx, raise, S((tn.name, S(nme.name)))), reverseRigid)
+          decls += nme.name -> PolymorphicType(thisCtx.lvl, bodyTy)
+          thisCtx.env += "." + nme.name -> 
+            PolymorphicType(thisCtx.lvl, FunctionType(subst(thisTy, reverseRigid), bodyTy)(prov))
+        }
+        td.mthDefs.foreach { case md @ MethodDef(rec, prt, nme, tparams, L(term)) =>
+          implicit val prov: TypeProvenance = tp(md.toLoc, "method definition")
+          val dummyTargs2 = tparams.map(p => freshVar(noProv/*TODO*/)(ctx.lvl + 2))
+          allTargsMaps(tn.name) += nme.name -> tparams.map(_.name).zip(dummyTargs2).toMap
+          if (!nme.name.head.isUpper)
+            err(msg"Method names must start with a capital letter", nme.toLoc)
+          if (defs.isDefinedAt(nme.name))
+            err(msg"Method '${tn}.${nme.name}' is already defined.", nme.toLoc)
+          val bodyTy = subst(typeLetRhs(rec, nme.name, term)(thisCtx, raise, S((tn.name, S(nme.name)))), reverseRigid)
+          defs += nme.name -> bodyTy
+          thisCtx.env += "." + nme.name -> 
+            PolymorphicType(bodyTy.level, FunctionType(subst(thisTy, reverseRigid), bodyTy.body)(prov))
+        }
+        newMthDecls += tn.name -> 
+          ((td.baseClasses.flatMap(t => newMthDecls.getOrElse(t.name, Map.empty))).toMap ++ decls.toMap)
+        newMthDefs += tn.name ->
+          ((td.baseClasses.flatMap(t => newMthDefs.getOrElse(t.name, Map.empty))).toMap ++ defs.toMap)
+      }
+      ctx.copy(mthDecls = newMthDecls.toMap, mthDefs = newMthDefs.toMap)
+    } |> { implicit ctx =>
+      val newEnv = MutMap.from(ctx.env)
+      val newMthDecls = ctx.mthDecls
+      val newMthDefs = ctx.mthDefs
+      newDefs.foreach { td =>
+        implicit val prov: TypeProvenance = tp(td.toLoc, "data definition")
+        val tn = td.nme
+        val decls = newMthDecls(tn.name)
+        val defs = newMthDefs(tn.name)
+        val rigidSubsMap = ctx.getTargsMaps(td.nme.name, "").getOrElse(Map.empty).collect { 
+            case s -> (tv: TypeVariable) => tv -> freshenAbove(ctx.lvl, tv, true)
```

Comment:
How can it be okay to filter out only the cases where the mapped value is a `TypeVariable`? That's very fishy. Couldn't it be a `ProxyType` at some point (maybe not now, but if the code base changes in the future)? Or an application `Id[X]` of the `type Id[T] = T` type? If we're indeed only interested in type variables here, this should be expressed in the type, to make sure we're not messing up, and for documentation.

Reply:
`targsMaps` store the mapping from type parameters to the dummy `TypeVariable`s, and to the rigidified `TraitTag`a when typing methods. It was originally just a map to `TypeVariable`s before I realized the class type parameters should be rigidified when typing methods.

I couldn't come up with a way to express this invariant in the type. I tried to use `TypeVariable \/ TraitTag`, but it would require a use of `asInstanceOf`, as `freshenAbove` returns a `SimpleType` even when called directly with a `TypeVariable`. For now I have added 3 `die` statements before I can come up with a better solution.

I think your worry about the other cases came from nested `TypeRef`s or `ProxyType`s in the body. They are in fact not handled by this map, but by the `subsMap` a few lines below. The `TypeVariable`s we are collecting are the dummy type arguments created when typing the body of the `TypeDef`. The actually applied type arguments are on the rhs of `subsMap`. I believe it is handling nested `AppliedType`s properly, as they would be typed as `TypeRef`s by `targsAppOf` (formerly `targsMapOf`. I am running out of names...).

However, as I was writing tests to test this out, I discovered that nested type arguments are not typed correctly, even in the `mlscript` branch:
```scala
class Foo[A]: { value: A }
class Bar[A]: Foo[Foo[A]]
//│ Defined class Foo
```
The `TypeDef` of `Bar` is not registered to the context, even though no error are emitted. I believe this is due to `checkRegular` returning `false`.

Reply:
Okay, thanks for the explanation. Now I understand what you are doing, although I don't really agree with how it's done. See my other comment https://github.com/hkust-taco/mlscript/pull/22#discussion_r757856454 for an idea to make the whole thing more functional, relying on computing the relevant information once, at the right place. We'd simply pass that relevant info along with each type definition being processed (instead of storing it in the context through the weird `targsMaps` mutable map), so that we can later perform subsumption checking after typing methods. Does that make sense to you?